### ISSUE TITLE
Cuda 12.1 - Fails to Build Here

### ISSUE THREAD

**Issue by ilikenwf** [turboderp/exllama#2]
```
exllama/exllama_ext/q4v2_matmul.cu(116): error: no instance of overloaded function "atomicAdd" matches the argument list
            argument types are: (half *, half)
      atomicAdd(out_.item_ptr(x_row, w_column), result);

```

**Comment by turboderp** [comment]
Are you sure you're on CUDA 12.1? Half-precision `atomicAdd` is supported by compute 7.x and up.

What output do you get from `nvcc --version`?

**Comment by ilikenwf** [comment]
Ah, my 1080ti may be too old to use it? I'm moderately new to this...

Was going to compare this to oobabooga with the current standard llama...with Wizard-Vicuna-13B-Uncensored-GPTQ and my 1080ti I'm getting around 16-17 tokens/s.

**Comment by turboderp** [comment]
Well, that version of `nvcc` should work, but Torch will still compile the extension for the compute capability of detected cards. And it looks like the 1080-Ti only supports compute 6.1, so it won't work with FP16.

At some point I do plan to add support for older hardware, but I'm not sure when I'll get to it.

**Comment by city96** [comment]
I've run into a similar issue with a different repo on my Pascal based Tesla P40 cards, which also have SM6.1. ( https://github.com/oobabooga/text-generation-webui/issues/1884#issuecomment-1541051291 )
The problem there was also the atomic add instruction for FP16.

The maintainer fixed it in the following commit: https://github.com/iwalton3/GPTQ-for-LLaMa/commit/59b71733af0f14ba9f3a7a81eb4e08a2cc772264

I'm not great with cpp / Cuda but with minimal changes this same code seems to make this project work on older cards as well. All code credit goes to @iwalton3

Added the following code to the top of `exllama_ext/q4v2_matmul.cu` (probably the wrong place but I just added it to the file that produced the error)

```cpp
#ifdef __CUDA_ARCH__
#if __CUDA_ARCH__ < 700
__device__ __forceinline__ void atomicAdd(half* address, half val) {
    unsigned int * address_as_ui =
      (unsigned int *) ((char *)address - ((size_t)address & 2));
    unsigned int old = *address_as_ui;
    unsigned int assumed;

  do {
    assumed = old;
    __half_raw hsum;
    hsum.x = (size_t)address & 2 ? (old >> 16) : (old & 0xffff);
    half tmpres = __hadd(hsum, val);
    hsum = __half_raw(tmpres);
    old = (size_t)address & 2 ? (old & 0xffff) | (hsum.x << 16) : (old & 0xffff0000) | hsum.x;
    old = atomicCAS(address_as_ui, assumed, old);
   } while (assumed != old);
}
#endif
#endif
```

**Comment by turboderp** [comment]
Thanks! I pushed an update. 

It's in a separate header and I refactored so I could test it on my 4090, though I still can't say if it will compile on a 1080 or if there are other compatibility issues. At least I can say it's doing a correct atomic addition. 

I've also only optimized for newer GPUs with way more cores and cache, so no idea what the performance will be like if it does work. At some point I do plan to test on and optimize for some older GPUs but I'm all out of PCIe slots right now. ;)

**Comment by city96** [comment]
It seems to work on my setup (also Cuda 12.1) so hopefully it also solves it for @ilikenwf !

As for the performance, it seems to be about the same, maybe a bit slower than the Cuda branch of GPTQ, though this is mainly because I'm heavily single-core CPU bound + as you said, probably don't benefit much from improvements aimed at newer GPU architectures either.

I could open a new issue on the CPU-bottleneck problem if you feel like looking into it, though the same thing happens with every other LLaMA code-base as well.
![CPU](https://github.com/turboderp/exllama/assets/125218114/92269b22-ea2b-4dbe-96e6-e2f48f750d51)

**Comment by turboderp** [comment]
Yeah, there's definitely a CPU bottleneck, most likely due to Python being quite slow, and due to Torch not actually being that efficient with small tensors. It looks like it trades a lot of overhead for the ability to deal efficiently with very large matmuls as you'd mostly encounter in training, but then during generation the hidden state ends up being a single row vector and the overhead per operation starts to matter.

I haven't actually profiled it to make sure, but if it's correct then the solution is to fuse more operations together, which I'm already working on.

**Comment by city96** [comment]
I've tried a few random things before (mostly torch ENV variables) but couldn't really track down the issue, glad you got some ideas for it!

**Comment by ilikenwf** [comment]
This does appear to build and run now!
```

-- Loading model
 -- Tokenizer: ../text-generation-webui/models/TheBloke_Wizard-Vicuna-13B-Uncensored-GPTQ/tokenizer.model
 -- Model config: ../text-generation-webui/models/TheBloke_Wizard-Vicuna-13B-Uncensored-GPTQ/config.json
 -- Model: ../text-generation-webui/models/TheBloke_Wizard-Vicuna-13B-Uncensored-GPTQ/Wizard-Vicuna-13B-Uncensored-GPTQ-4bit-128g.compat.no-act-order.safetensors
 -- Sequence length: 2048
 -- Options: ['attention: pytorch_scaled_dp', 'matmul: switched']
 ** Time, Load model: 12.58 seconds
 -- Groupsize (inferred): 128
 -- Act-order (inferred): no
 ** VRAM, Model: [cuda:0] 6,996.23 MB

```

**Comment by ilikenwf** [comment]
Chatbot throws an error, lol... 
```

exllama/cuda_ext.py", line 211, in llama_rms_norm
    rms_norm(x, w, output, scratch, epsilon)
RuntimeError: w must be a half tensor

```

**Comment by turboderp** [comment]
Yes, I've not had time to test it with every GPTQ model out there, and there are some that use different datatypes in places because the format has been changing so much. A big part of the motivation for this project is to create something a little more stable to build on, after I got fed up with the constant, breaking changes to GPTQ-for-LLaMa. :)

Anyway, I'll have a look at that particular conversion of Vicuna later today, but in the meantime you could try changing line 386 of model.py from this:

        self.weight = tensors[key]

To this:

        self.weight = tensors[key].half()

Which should force all layer norm weights to the correct datatype.

**Comment by 0cc4m** [comment]
@turboderp I cannot build the exllama cuda code for Pascal GPUs (Compute capability 6.0/6.2) because it cannot distinguish the half2 overload. Do you know a solution for this?

```
FAILED: /home/user/exllama/build/temp.linux-x86_64-cpython-311/exllama_ext/cuda_func/q4v2_matmul.o
/opt/cuda/bin/nvcc  -I/home/user/exllama/venv/lib/python3.11/site-packages/torch/include -I/home/user/exllama/venv/lib/python3.11/site-packages/torch/include/torch/csrc/api/include -I/home/user/exllama/venv/lib/python3.11/site-packages/torch/include/TH -I/home/user/exllama/venv/lib/p
ython3.11/site-packages/torch/include/THC -I/opt/cuda/include -I/home/user/exllama/venv/include -I/usr/include/python3.11 -c -c /home/user/exllama/exllama_ext/cuda_func/q4v2_matmul.cu -o /home/user/exllama/build/temp.linux-x86_64-cpython-311/exllama_ext/cuda_func/q4v2_matmul.o -D__CU
DA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr --compiler-options ''"'"'-fPIC'"'"'' -O3 -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"'
'-DPYBIND11_BUILD_ABI="_cxxabi1011"' -DTORCH_EXTENSION_NAME=exllama_ext -D_GLIBCXX_USE_CXX11_ABI=0 -gencode=arch=compute_60,code=sm_60 -gencode=arch=compute_61,code=sm_61 -gencode=arch=compute_86,code=compute_86 -gencode=arch=compute_86,code=sm_86 -std=c++17
/home/user/exllama/exllama_ext/cuda_func/../cuda_compat.cuh(48): error: cannot overload functions distinguished by return type alone
  __attribute__((device)) __inline__ __attribute__((always_inline)) void atomicAdd(half2* address, half2 val) { atomicAdd_half2(address, val); }
  ```

Python 3.11, CUDA 12.1

**Comment by turboderp** [comment]
I need to do a little bit more research to figure out exactly what's supported on the various, many compute versions. But for now, you can try changing the line:

    #if __CUDA_ARCH__ < 700

To something else, like:

   #if __CUDA_ARCH__ < 600

In `exllama_ext/cuda_compat.cuh`.

**Comment by turboderp** [comment]
I'm confused. So it provides `atomicAdd()` for `half2` but not for `half`?

**Comment by 0cc4m** [comment]
No, my guess is it can't distinguish between half and half2, so the compiler can't overload the function like you did here. I got it to work by replacing the calls to atomicAdd with explicit atomicAdd_half and half2 calls. Not yet sure why this is happening.

**Comment by bkutasi** [comment]
Hey there, glad to see others face the same problem. Adding .half() into model line 386 didn't help(i found it on line 240), but adding the following two things helped:

> I need to do a little bit more research to figure out exactly what's supported on the various, many compute versions. But for now, you can try changing the line:
> 
> ```
> #if __CUDA_ARCH__ < 700
> ```
> 
> To something else, like:
> 
> #if **CUDA_ARCH** < 600
> 
> In `exllama_ext/cuda_compat.cuh`.

And this one:

> I've run into a similar issue with a different repo on my Pascal based Tesla P40 cards, which also have SM_61. ( [oobabooga/text-generation-webui#1884 (comment)](https://github.com/oobabooga/text-generation-webui/issues/1884#issuecomment-1541051291) ) The problem there was also the atomic add instruction for FP16.
> 
> The maintainer fixed it in the following commit: [iwalton3/GPTQ-for-LLaMa@59b7173](https://github.com/iwalton3/GPTQ-for-LLaMa/commit/59b71733af0f14ba9f3a7a81eb4e08a2cc772264)
> 
> I'm not great with cpp / Cuda but with minimal changes this same code seems to make this project work on older cards as well. All code credit goes to @iwalton3
> 
> Added the following code to the top of `exllama_ext/q4v2_matmul.cu` (probably the wrong place but I just added it to the file that produced the error)
> 
> ```c++
> #ifdef __CUDA_ARCH__
> #if __CUDA_ARCH__ < 700
> __device__ __forceinline__ void atomicAdd(half* address, half val) {
>     unsigned int * address_as_ui =
>       (unsigned int *) ((char *)address - ((size_t)address & 2));
>     unsigned int old = *address_as_ui;
>     unsigned int assumed;
> 
>   do {
>     assumed = old;
>     __half_raw hsum;
>     hsum.x = (size_t)address & 2 ? (old >> 16) : (old & 0xffff);
>     half tmpres = __hadd(hsum, val);
>     hsum = __half_raw(tmpres);
>     old = (size_t)address & 2 ? (old & 0xffff) | (hsum.x << 16) : (old & 0xffff0000) | hsum.x;
>     old = atomicCAS(address_as_ui, assumed, old);
>    } while (assumed != old);
> }
> #endif
> #endif
> ```

I have a 1080ti and a 3080ti, but this seems to be only running on the 3080ti, but maybe it gets confused with the 1080ti in the system. I wonder if i can force it to run on 1080ti? I also played around with the driver before so that could also cause an issue but other programs like llama.ccp with cublas worked without a problem.

**Comment by turboderp** [comment]
I've added a new option so you can run it without using half2s anywhere. I'm not sure if that helps on the 1080, but you can try `-nh2` to see what happens. It might still give compilation errors, but it shouldn't care if you replace atomicAdd_half2 with a dummy function. Something to experiment with, at least, idk. :)

**Comment by 0cc4m** [comment]
I managed to work around the compile issue like [this](https://github.com/0cc4m/exllama/commit/63653e56ce5cef1d76a5160dbf0d61bb9fb40265). Not a real fix, but it works for now.

**Comment by bkutasi** [comment]
Yep, the same `-gencode=arch=compute_61,code=sm_61 -gencode=arch=compute_86,code=compute_86 -gencode=arch=compute_86,code=sm_86`

**Comment by 0cc4m** [comment]
In my case setting `CUDA_VISIBLE_DEVICES` to not show Pascal worked.

**Comment by bkutasi** [comment]
> I don't know what to make of that, but it clearly thinks is still needs to generate code for Pascal. Maybe it suggests there's something up with your env that could also help to explain the half2 stuff?

I will go through the code sooner or later any try to figure it out. If others have no problem, it's probably my env messed up. For the meantime I will just unplug my pascal card.

**Comment by turboderp** [comment]
I'll consider this resolved for now, just to get some issues closed, but feel free to reopen if necessary.
